flink-test-utils-parent/flink-connector-testing-framework/src/main/java/org/apache/flink/connectors/test/common/testsuites/TestSuiteBase.java

```diff
@@ -0,0 +1,430 @@
+
+        // Check test result
+        checkMultipleSplitRecords(
+                testRecordCollections.stream()
+                        .map(Collection::iterator)
+                        .collect(Collectors.toList()),
+                resultIterator);
+        resultIterator.close();
+    }
+
+    /**
+     * Test connector source with a redundant parallelism.
+     *
+     * <p>This test will create 4 split in the external system, write test data to all splits, and
+     * consume back via a Flink job with 5 parallelism, so at least one parallelism / source reader
+     * will be idle (assigned with no splits). If the split enumerator of the source doesn't signal
+     * NoMoreSplitsEvent to the idle source reader, the Flink job will never spin to FINISHED state.
+     *
+     * <p>The number and order of records in each split consumed by Flink need to be identical to
+     * the test data written into the external system to pass this test. There's no requirement for
+     * record order across splits.
+     *
+     * <p>A bounded source is required for this test.
+     */
+    @Case
+    @DisplayName("Test source with at least one idle parallelism")
+    public void testRedundantParallelism(
+            TestEnvironment testEnv, ExternalContext<T> externalContext) throws Exception {
+
+        final int splitNumber = 4;
+        final List<Collection<T>> testRecordCollections = new ArrayList<>();
+        for (int i = 0; i < splitNumber; i++) {
+            testRecordCollections.add(generateAndWriteTestData(externalContext));
+        }
+
+        final CloseableIterator<T> resultIterator =
+                testEnv.createExecutionEnvironment()
+                        .fromSource(
+                                externalContext.createSource(Boundedness.BOUNDED),
+                                WatermarkStrategy.noWatermarks(),
+                                "Tested Source")
+                        .setParallelism(splitNumber + 1)
+                        .executeAndCollect("Redundant Parallelism Test");
+
+        checkMultipleSplitRecords(
+                testRecordCollections.stream()
+                        .map(Collection::iterator)
+                        .collect(Collectors.toList()),
+                resultIterator);
+    }
+
+    /**
+     * Test connector source with task manager failover.
+     *
+     * <p>This test will create 1 split in the external system, write test record set A into the
+     * split, restart task manager to trigger job failover, write test record set B into the split,
+     * and terminate the Flink job finally.
+     *
+     * <p>The number and order of records consumed by Flink should be identical to A before the
+     * failover and B after the failover in order to pass the test.
+     *
+     * <p>An unbounded source is required for this test, since TaskManager failover will be
+     * triggered in the middle of the test.
```

Comment:
We probably should add a way for users to explicitly disable test on bounded/unbounded jobs by a simple annotation. (e.g. a source doesn't support different boundedness)